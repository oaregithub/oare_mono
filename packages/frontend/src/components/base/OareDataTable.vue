<template>
  <v-data-table
    v-bind="$attrs"
    :options.sync="sortOptions"
    :loading="loading"
    :items="loading ? [] : $attrs.items"
    :footer-props="{
      'items-per-page-options': [10, 25, 50, 100],
    }"
  >
    <template
      v-for="(_, scopedSlotName) in $scopedSlots"
      v-slot:[scopedSlotName]="slotData"
    >
      <slot :name="scopedSlotName" v-bind="slotData" />
    </template>

    <template v-for="(_, slotName) in $slots" v-slot:[slotName]>
      <slot :name="slotName" />
    </template>
  </v-data-table>
</template>

<script lang="ts">
import {
  defineComponent,
  ref,
  PropType,
  watch,
  computed,
} from '@vue/composition-api';
import useQueryParam from '@/hooks/useQueryParam';
import { DataOptions } from 'vuetify';
import sl from '@/serviceLocator';

export interface OareDataTableOptions {
  page: number;
  rows: number;
  sortBy: string;
  sortDesc: boolean;
}

export default defineComponent({
  props: {
    debounce: {
      type: Boolean,
      default: false,
    },
    defaultSort: {
      type: String,
      default: '',
    },
    defaultDesc: {
      type: Boolean,
      default: true,
    },
    defaultPage: {
      type: Number,
      default: 1,
    },
    defaultRows: {
      type: Number,
      default: 10,
    },
    errorMessage: {
      type: String,
      default: 'Failed to fetch items',
    },
    fetchItems: {
      type: Function as PropType<
        (options: OareDataTableOptions) => Promise<void>
      >,
      required: true,
    },
    watchedParams: {
      type: Array as PropType<string[]>,
      default: () => [],
    },
  },
  setup({
    debounce,
    defaultSort,
    defaultDesc,
    defaultPage,
    defaultRows,
    fetchItems,
    errorMessage,
    watchedParams,
  }) {
    const _ = sl.get('lodash');
    const actions = sl.get('globalActions');
    const sortBy = useQueryParam('sortBy', defaultSort);
    const sortDesc = useQueryParam('sortDesc', String(defaultDesc));
    const page = useQueryParam('page', String(defaultPage));
    const rows = useQueryParam('rows', String(defaultRows));

    const loading = ref(false);

    const sortOptions = ref<DataOptions>({
      page: Number(page.value),
      itemsPerPage: Number(rows.value),
      sortBy: [sortBy.value],
      sortDesc: [sortDesc.value === 'true'],
      groupBy: [],
      groupDesc: [],
      multiSort: false,
      mustSort: true,
    });

    const tableOptions = computed<OareDataTableOptions>(() => ({
      page: Number(page.value),
      rows: Number(rows.value),
      sortBy: sortBy.value,
      sortDesc: sortDesc.value === 'true',
    }));

    const queryParams = watchedParams.map(paramName =>
      useQueryParam(paramName, '')
    );

    const performFetch = async () => {
      try {
        loading.value = true;
        await fetchItems(tableOptions.value);
      } catch (err) {
        actions.showErrorSnackbar(errorMessage, err as Error);
      } finally {
        loading.value = false;
      }
    };

    const queryParamsOnChange = () => {
      sortOptions.value.page = 1;
      performFetch();
    };

    watch(
      queryParams,
      debounce ? _.debounce(queryParamsOnChange, 500) : queryParamsOnChange
    );

    watch(sortOptions, newOptions => {
      sortBy.value = newOptions.sortBy[0];
      sortDesc.value = String(newOptions.sortDesc[0]);
      page.value = String(newOptions.page);
      rows.value = String(newOptions.itemsPerPage);

      performFetch();
    });

    return {
      sortOptions,
      loading,
    };
  },
});
</script>
