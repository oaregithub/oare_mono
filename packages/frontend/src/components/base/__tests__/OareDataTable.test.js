import { mount, createLocalVue } from '@vue/test-utils';
import Vuetify from 'vuetify';
import VueCompositionApi from '@vue/composition-api';
import flushPromises from 'flush-promises';
import sl from '@/serviceLocator';
import OareDataTable from '../OareDataTable.vue';

const vuetify = new Vuetify();
const localVue = createLocalVue();
localVue.use(VueCompositionApi);

describe('OareDataTable', () => {
  const headers = [
    {
      text: 'Name',
      value: 'name',
    },
    {
      text: 'Age',
      value: 'age',
    },
  ];

  const items = [
    {
      name: 'Jeff',
      age: 15,
    },
    {
      name: 'Amy',
      age: 22,
    },
  ];

  const actions = {
    showErrorSnackbar: jest.fn(),
  };

  const fetchItems = jest.fn().mockResolvedValue();

  const createWrapper = (props = {}) =>
    mount(OareDataTable, {
      vuetify,
      localVue,
      propsData: {
        headers,
        items,
        defaultSort: 'name',
        fetchItems,
        ...props,
      },
    });

  beforeEach(() => {
    sl.set('globalActions', actions);
  });

  it('should render', () => {
    const wrapper = createWrapper();

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('fetches items when performing sort', async () => {
    const wrapper = createWrapper();
    await wrapper.findAll('th').at(0).trigger('click');
    await flushPromises();

    expect(fetchItems).toHaveBeenCalled();
  });

  it('shows error if fetching items fails', async () => {
    const errorMessage = 'Fetching items failed';
    createWrapper({
      errorMessage,
      fetchItems: jest.fn().mockRejectedValue('could not fetch items'),
    });

    await flushPromises();
    expect(actions.showErrorSnackbar).toHaveBeenCalled();
  });
});
