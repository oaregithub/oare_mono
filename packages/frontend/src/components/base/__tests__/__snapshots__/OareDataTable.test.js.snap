// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OareDataTable should render 1`] = `
<div class="v-data-table v-data-table--has-bottom theme--light" headers="[object Object],[object Object]" items="[object Object],[object Object]">
  <div class="v-data-table__wrapper">
    <table>
      <colgroup>
        <col class="">
        <col class="">
      </colgroup>
      <thead class="v-data-table-header">
        <tr>
          <th role="columnheader" scope="col" aria-label="Name: Sorted descending. Activate to sort ascending." aria-sort="descending" class="text-start sortable active desc"><span>Name</span><i aria-hidden="true" class="v-icon notranslate v-data-table-header__icon mdi mdi-arrow-up theme--light" style="font-size: 18px;"></i></th>
          <th role="columnheader" scope="col" aria-label="Age: Not sorted. Activate to sort ascending." aria-sort="none" class="text-start sortable"><span>Age</span><i aria-hidden="true" class="v-icon notranslate v-data-table-header__icon mdi mdi-arrow-up theme--light" style="font-size: 18px;"></i></th>
        </tr>
      </thead>
      <tbody>
        <tr class="">
          <td class="text-start">Jeff</td>
          <td class="text-start">15</td>
        </tr>
        <tr class="">
          <td class="text-start">Amy</td>
          <td class="text-start">22</td>
        </tr>
      </tbody>
    </table>
  </div>
  <div class="v-data-footer">
    <div class="v-data-footer__select">Rows per page:<div class="v-input v-input--hide-details v-input--is-label-active v-input--is-dirty theme--light v-text-field v-select">
        <div class="v-input__control">
          <div role="button" aria-haspopup="listbox" aria-expanded="false" aria-owns="list-10" class="v-input__slot">
            <div class="v-select__slot">
              <div class="v-select__selections">
                <div class="v-select__selection v-select__selection--comma">10</div><input aria-label="Rows per page:" id="input-10" readonly="readonly" type="text" aria-readonly="false" autocomplete="off">
              </div>
              <div class="v-input__append-inner">
                <div class="v-input__icon v-input__icon--append"><i aria-hidden="true" class="v-icon notranslate mdi mdi-menu-down theme--light"></i></div>
              </div><input type="hidden" value="10">
            </div>
            <div class="v-menu">
              <!---->
            </div>
          </div>
        </div>
      </div>
    </div>
    <div class="v-data-footer__pagination">1-2 of 2</div>
    <div class="v-data-footer__icons-before"><button type="button" disabled="disabled" class="v-btn v-btn--disabled v-btn--icon v-btn--round v-btn--text theme--light v-size--default" aria-label="Previous page"><span class="v-btn__content"><i aria-hidden="true" class="v-icon notranslate mdi mdi-chevron-left theme--light"></i></span></button></div>
    <div class="v-data-footer__icons-after"><button type="button" disabled="disabled" class="v-btn v-btn--disabled v-btn--icon v-btn--round v-btn--text theme--light v-size--default" aria-label="Next page"><span class="v-btn__content"><i aria-hidden="true" class="v-icon notranslate mdi mdi-chevron-right theme--light"></i></span></button></div>
  </div>
</div>
`;
